### PR TITLE
Edge analysis improvements

### PR DESCRIPTION
**Pull request by mjohnson541**
This has a set of assorted improvements primarily related to edge analysis. These improvements broadly facilitate integration of RMS inside of RMG and this PR is necessary for the companion PR (soon to be created) on RMG-Py. Probably the most significant are the full deprecation of Species.index and enabling of model resurrection.

### CODE REVIEW

File: src/Domain.jl [hunk 1]
```diff
@@ -96,7 +96,7 @@ function ConstantTPDomain(;phase::E2,initialconds::Dict{X,X2},constantspecies::A
         jacobian=zeros(typeof(T),length(phase.species),length(phase.species))
     end
     rxnarray = getreactionindices(phase)
-    return ConstantTPDomain(phase,[phase.species[1].index,phase.species[end].index,phase.species[end].index+1],[1,length(phase.species)+length(phase.reactions)],constspcinds,
+    return ConstantTPDomain(phase,[1,length(phase.species),length(phase.species)+1],[1,length(phase.species)+length(phase.reactions)],constspcinds,
         T,P,kfs,krevs,efficiencyinds,Gs,rxnarray,mu,diffs,jacobian,sensitivity,false,MVector(false),MVector(0.0),p, Dict("V"=>phase.species[end].index+1)), y0, p
```
Comment (hwpang) [inline src/Domain.jl:100]: The species index is also used in the thermovariabledict, maybe change `Dict("V"=>phase.species[end].index+1))` to `Dict("V"=>length(phase.species)+1))`? Same for following domains.

File: src/Domain.jl [hunk 2]
```diff
@@ -313,7 +313,7 @@ function ParametrizedTPDomain(;phase::Z,initialconds::Dict{X,Any},constantspecie
         jacobian=zeros(typeof(V),length(phase.species)+1,length(phase.species)+1)
```
Comment (hwpang) [inline src/Domain.jl:313]: I'm not able to comment at the exact position due to github limitation.. but species index is also used in line 300 and line 608 (`y0[phase.species[1].index:phase.species[end].index] = ns`), maybe change them as well for consistency.

File: src/Domain.jl [hunk 3]
```diff
@@ -96,7 +96,7 @@ function ConstantTPDomain(;phase::E2,initialconds::Dict{X,X2},constantspecies::A
         jacobian=zeros(typeof(T),length(phase.species),length(phase.species))
     end
     rxnarray = getreactionindices(phase)
-    return ConstantTPDomain(phase,[phase.species[1].index,phase.species[end].index,phase.species[end].index+1],[1,length(phase.species)+length(phase.reactions)],constspcinds,
+    return ConstantTPDomain(phase,[1,length(phase.species),length(phase.species)+1],[1,length(phase.species)+length(phase.reactions)],constspcinds,
```
Comment (hwpang) [inline src/Domain.jl:99]: Is this change for dealing with the coreedgedomain cases where the species number could change or where the index doesn't match with the species's exact position in the array? Could you give more context on this change?

File: src/Phase.jl [hunk 4]
```diff
@@ -79,15 +79,16 @@ function IdealDiluteSolution(species,reactions,solvent; name="",diffusionlimited
...
     if all(echangevec .== 0)
         electronchange = nothing
     else 
         electronchange = convert(echangevec,Array{Float64,1})
     end
     reversibility = getfield.(rxns,:reversible)
-    rxnarray = getreactionindices(species,rxns)
+
     return IdealDiluteSolution(species=species,reactions=rxns,solvent=solvent,name=name,
         spcdict=Dict([sp.name=>sp.index for sp in species]),stoichmatrix=M,Nrp=Nrp,rxnarray=rxnarray,veckinetics=vectuple,
```
Comment (hwpang) [inline src/Phase.jl:93]: Should this be `Dict([sp.name=>i for (i,sp) in enumerate(species)])` if we are avoiding to use the species' index to indicate the species's position in an array?

**Review by hwpang — CHANGES_REQUESTED** [review]
Thanks for the PR! Mostly looks good to me. Only a few questions/comments

File: src/Phase.jl (hunk 4 above)
Comment (mjohnson541) [reply src/Phase.jl]: fixed

File: src/Domain.jl (hunk 2 above)
Comment (mjohnson541) [reply src/Domain.jl:313]: fixed

File: src/Domain.jl (hunk 1 above)
Comment (mjohnson541) [reply src/Domain.jl]: fixed

**Comment by mjohnson541** [comment]
Okay I think I've fixed everything and tests pass!

**Review by hwpang — APPROVED** [review]
Thanks for the changes! Looks good to me!

**Pull request merged**